plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/router/MikamiTabuchiConnectionRouter.java

```diff
@@ -0,0 +1,284 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.erd.ui.router;
+
+
+import org.eclipse.draw2dl.*;
+import org.eclipse.draw2dl.LayoutListener.Stub;
+import org.eclipse.draw2dl.geometry.Point;
+import org.eclipse.draw2dl.geometry.PointList;
+import org.eclipse.draw2dl.geometry.PrecisionPoint;
+import org.eclipse.draw2dl.geometry.Rectangle;
+import org.jkiss.dbeaver.ui.UIUtils;
+
+import java.util.*;
+
+
+public class MikamiTabuchiConnectionRouter extends AbstractRouter {
+    private boolean isDirty = false;
+    private final Map<Connection, Object> constraintMap = new HashMap<>();
+    private Map<IFigure, Rectangle> figuresToBounds;
+    private Map<Connection, OrthogonalPath> connectionToPath;
+    private final IFigure container;
+    private final MikamiTabuchiRouter algorithm = new MikamiTabuchiRouter();
+    private final Set<Connection> staleConnections = new HashSet<>();
+    private boolean ignoreInvalidate;
+    private final LayoutListener listener = new MikamiTabuchiConnectionRouter.LayoutTracker();
+    private final FigureListener figureListener = source -> {
+        Rectangle newBounds = source.getBounds().getCopy();
+        if (MikamiTabuchiConnectionRouter.this.algorithm.updateObstacle(MikamiTabuchiConnectionRouter.this.figuresToBounds.get(source), newBounds)) {
+            MikamiTabuchiConnectionRouter.this.queueRerouting();
+            MikamiTabuchiConnectionRouter.this.isDirty = true;
+        }
+
+        MikamiTabuchiConnectionRouter.this.figuresToBounds.put(source, newBounds);
+    };
+
+
+    public Object getConstraint(Connection connection) {
+        return this.constraintMap.get(connection);
+    }
+
+
+    public void invalidate(Connection connection) {
+        if (!this.ignoreInvalidate) {
+            this.staleConnections.add(connection);
+            this.isDirty = true;
+        }
+    }
+
+    private void queueRerouting() {
+        if (this.connectionToPath != null && !this.connectionToPath.isEmpty()) {
+            try {
+                this.ignoreInvalidate = true;
+                ((Connection)this.connectionToPath.keySet().iterator().next()).revalidate();
+            } finally {
+                this.ignoreInvalidate = false;
+            }
+        }
+    }
+
+    void addChild(IFigure child) {
+        if (this.connectionToPath != null && !this.figuresToBounds.containsKey(child)) {
+            Rectangle bounds = child.getBounds().getCopy();
+            this.algorithm.addObstacle(bounds);
+            this.figuresToBounds.put(child, bounds);
+            child.addFigureListener(this.figureListener);
+            this.isDirty = true;
+
+        }
+    }
+
+    public MikamiTabuchiConnectionRouter(IFigure container) {
+        this.container = container;
+    }
```

Comment:
Why is the constructor appears here? It must be located at the very top just after field declarations.